internal/cluster/distribution/pke/pkeaws/pkeawsworkflow/activity_nlb_health_check.go

```diff
@@ -0,0 +1,143 @@
+		awsSessionFactory: awsSessionFactory,
+		elbv2Factory:      elbv2Factory,
+	}
+}
+
+type HealthCheckActivityInput struct {
+	OrganizationID uint
+	SecretID       string
+	Region         string
+	ClusterID      uint
+	ClusterName    string
+}
+
+// HealthCheckActivityOutput holds the output data
+type HealthCheckActivityOutput struct {
+	nlbStatus []*elbv2.TargetHealthDescription
+}
+
+// Register registers the activity in the worker.
+func (a *HealthCheckActivity) Register() {
+	activity.RegisterWithOptions(a.Execute, activity.RegisterOptions{Name: HealthCheckActivityName})
+}
+
+func (a *HealthCheckActivity) Execute(ctx context.Context, input HealthCheckActivityInput) (*HealthCheckActivityOutput, error) {
+	if a == nil {
+		return nil, errors.New("activity is nil")
+	}
+	session, err := a.awsSessionFactory.New(input.OrganizationID, input.SecretID, input.Region)
+	if err != nil {
+		return nil, errors.WrapIf(err, "failed to create AWS session")
+	}
+
+	elbv2Svc := a.elbv2Factory.New(session)
+	nlbName := "pke-" + input.ClusterName + "-nlb"
+
+	backoffConfig := backoff.ConstantBackoffConfig{
+		Delay:      backoffDelay,
+		MaxRetries: backoffMaxRetries,
+	}
+	backoffPolicy := backoff.NewConstantBackoffPolicy(backoffConfig)
+
+	var loadBalancerOutput *elbv2.DescribeLoadBalancersOutput
+	if err := backoff.Retry(func() error {
+		var errInt error
+		loadBalancerOutput, errInt = elbv2Svc.DescribeLoadBalancers(
+			&elbv2.DescribeLoadBalancersInput{
+				Names: []*string{&nlbName},
+			})
+		if errInt != nil {
+			return errors.WrapIf(errInt, "failed to describe load balancers")
+		}
+
+		return nil
+	}, backoffPolicy); err != nil {
+		return nil, err
+	}
+
+	var targetGroupOutput *elbv2.DescribeTargetGroupsOutput
+	if err := backoff.Retry(func() error {
+		var errInt error
+		targetGroupOutput, errInt = elbv2Svc.DescribeTargetGroups(
+			&elbv2.DescribeTargetGroupsInput{
+				LoadBalancerArn: loadBalancerOutput.LoadBalancers[0].LoadBalancerArn,
+			})
+		if errInt != nil {
+			return errors.WrapIf(errInt, "failed to describe target groups")
+		}
+
+		return nil
+	}, backoffPolicy); err != nil {
+		return nil, err
+	}
+
+	var targetHealthOutput *elbv2.DescribeTargetHealthOutput
+	if err := backoff.Retry(func() error {
+		var errInt error
+		targetHealthOutput, errInt = elbv2Svc.DescribeTargetHealth(
+			&elbv2.DescribeTargetHealthInput{
+				TargetGroupArn: targetGroupOutput.TargetGroups[0].TargetGroupArn,
+			})
+		if errInt != nil {
+			return errors.WrapIf(errInt, "failed to describe target health")
+		}
+
+		return nil
+	}, backoffPolicy); err != nil {
+		return nil, err
+	}
+
+	for _, tgHealhtDesc := range targetHealthOutput.TargetHealthDescriptions {
+		if *tgHealhtDesc.TargetHealth.State != elbv2.TargetHealthStateEnumHealthy {
+			return &HealthCheckActivityOutput{
+				nlbStatus: targetHealthOutput.TargetHealthDescriptions,
+			}, errors.New("not all targets of the network load balancer are healthy")
+		}
+	}
+
+	return nil, nil
```

Comment:
**Optional**: we **MAY** consider the more idiomatic approach of either returning a successful value with nil error or a nil value with an error, but never both.

Could we incorporate the returned value into the error content as details or would we loose any information then?
As I can see in success case we return <nil, nil>.

Reply:
fixed: https://github.com/banzaicloud/pipeline/pull/3229/commits/6146e54ad33a4aa0d22fad813f9c67841de394ea

Reply:
You only partially addressed what I commented.

In the error case we are still returning <non-nil-value, error message> which is not really idiomatic in Golang.
Is this intentional and necessary?

Reply:
This is also important, because I read it somewhere in the Cadence documentation/code that they propagate nil value in case of error so I believe the non-nil return value is thrown away anyway, trying to find the source of that information again.

Meaning these

```golang
			return &HealthCheckActivityOutput{
				nlbStatus: targetHealthOutput.TargetHealthDescriptions,
			}, errors.NewWithDetails("unable to getting state of the target health description", "targetHealthDesc", tgHealthDesc)
...
			return &HealthCheckActivityOutput{
				nlbStatus: targetHealthOutput.TargetHealthDescriptions,
			}, errors.NewWithDetails("not all targets of the network load balancer are healthy", "targetHealthDesc", tgHealthDesc)
```

Can't we do something like this:

```golang
return nil, errors.NewWithDetails("<message>", "targetHealthDescriptions", targetHealthOutput.TargetHealthDescriptions, "targetHealthDesc", tgHealthDesc)
```

Reply:
Nevermind, based on the [0.13.4 implementation](https://github.com/uber-go/cadence-client/blob/cadef60d6bd144a8b36c1c555bbf559273e9b494/internal/workflow.go#L462) it seems like the return values get propagated in case of error as well.

Reply:
fixed: https://github.com/banzaicloud/pipeline/pull/3229/commits/12c03d74786f9c865b56632b3481ce203a09b584